internal/search/crawler/crawler.go

```diff
@@ -0,0 +1,62 @@
+package crawler
+
+import (
+	"context"
+	"log"
+	"os"
+	"sync"
+
+	"github.com/damienr74/kustomize/internal/search/doc"
+)
+
+// The crawler.Interface retrieves doc.KustomizationDocument
+type Interface interface {
+	// Crawl returns when it is done processing. This method does not take ownership of
+	// the channel.
+	Crawl(context.Context, chan<- *doc.KustomizationDocument) error
+}
+
+// This function uses the output channel to forward kustomization documents from a list of crawlers.
+// The output can be consumed by an
+//
+// The return value is an array of errors in which each index represents the index of the crawler that emitted
+// the error. Although the errors themselves can be nil, the array will always be exactly the size of the
+// crawlers array.
+//
+// Crawl is a blocking function and only returns once all of the crawlers are finished with execution.
+func Crawl(ctx context.Context, output chan<- *doc.KustomizationDocument, crawlers []Interface) []error {
+	logger := log.New(os.Stdin, "Crawl:", log.LUTC|log.Ltime|log.Ldate)
+
+	errs := make([]error, len(crawlers))
+
+	wg := sync.WaitGroup{}
+
+	for i, crawler := range crawlers {
+		// Crawler implementations get their own channels to prevent one crawler from closing the output stream.
+		docs := make(chan *doc.KustomizationDocument)
+		wg.Add(2)
+
+		// Forward all of the documents from this crawler's channel to the main output channel.
+		go func(docs <-chan *doc.KustomizationDocument) {
+			defer wg.Done()
+			for doc := range docs {
+				output <- doc
+			}
+		}(docs)
+
+		// Run this crawler and capture it's returned error.
+		go func(idx int, crawler Interface, docs chan<- *doc.KustomizationDocument) {
+			defer func() {
+				wg.Done()
+				if r := recover(); r != nil {
+					logger.Printf("%+v panicked: %v", crawler, r)
+				}
+			}()
+			defer close(docs)
+			errs[idx] = crawler.Crawl(ctx, docs)
```

Comment:
The recover case is required in case the crawler erroneously closes the channel allocated for it. We wouldn't want to make all of the crawlers fail just because one of them does.

What should happen to the value in `errs[idx]`? I feel like there should be some type of way to check that panics happened from the call site of the `Crawl` method, but wasn't sure on what that should look like.

At the very least, it could wrap any existing errors and add the panic value to the error. The crawler method could also return another value for panics, or it could be left as is. Feel free to suggest anything else, just wanted to bring it to your attention.

Reply:
I guess another way to implement this would be to have the `Crawl` method handle only a single crawler, but then this delegates the responsibility of not making everything crash to the caller, which I don't really think is a good idea.

Reply:
its fine to have a CrawlerRunner, MegaCrawler, etc.  Define a Crawler well, then define a thing that manages N of them.

agree that one crawl failure shouldn't fail them all.
errors can just be accumulated and returned in one error.  refine later, don't sweat it.